### DISCUSSION TITLE
HTTP Basic auth does not correctly decode passwords if e.g. "Ä" or some similar characters are included

### DISCUSSION THREAD

**Discussion by TheCataliasTNT2k** [fastapi/fastapi#8856] [Questions]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
import uvicorn
from fastapi import FastAPI, Depends, HTTPException
from fastapi.security import HTTPBasicCredentials, HTTPBasic
from starlette import status

app = FastAPI()

async def auth(creds: HTTPBasicCredentials = Depends(HTTPBasic())):
    username = creds.username
    password = creds.password
    print(password)
    if password == "test" or password == "ä":
        return "OK"  # will not be returned, when bug occurs

    raise HTTPException(
        status_code=status.HTTP_401_UNAUTHORIZED,
        detail="Incorrect email or password",
        headers={"WWW-Authenticate": "Basic"},
    )

@app.get("/", dependencies=[Depends(auth)])
def read_root():
    return {"Hello": "World"}

uvicorn.run(f"{__name__}:app", host="0.0.0.0", port=5000, reload=False)
```

### Description

run following in linux terminal (or open page in browser, reset session between calls (private window))
syntax: `command # expected response`
```
curl -u username:password http://localhost:5000; echo  # should be '{"detail":"Invalid authentication credentials"}'
curl -u username:test http://localhost:5000; echo  # should be '{"Hello":"World"}'
curl -u username:ä http://localhost:5000; echo  # should be '{"Hello":"World"}'
```

### Environment

* OS: [e.g. Linux / Windows / macOS]: Linux
* FastAPI Version [e.g. 0.3.0]: 0.65.1

To know the FastAPI version use:

```bash
python -c "import fastapi; print(fastapi.__version__)"
```

* Python version: 3.9.4

To know the Python version use:

```bash
python --version
```

### Additional context

![image](https://user-images.githubusercontent.com/44349750/118381496-b8f5f180-b5eb-11eb-8355-0cd554f6683a.png)

The error is caused by the decode call after b64decode, because "ascii" is used as encoding:
```Python
try:
    data = b64decode(param).decode("ascii")
except (ValueError, UnicodeDecodeError, binascii.Error):
     raise invalid_user_credentials_exc
```
Can be found [here](https://github.com/tiangolo/fastapi/blob/43df5d05edd30fdeb16a9f5b26c02d4f47bdd8cb/fastapi/security/http.py#L85)

**Comment by HarrySky** [comment]
I am having the same issue, solved it by using custom `HTTPBasic` class that uses UTF-8 instead of ASCII (and I removed some checks I don't need):

```python
from base64 import b64decode
from typing import Optional

from fastapi import HTTPException, status
from fastapi.security import HTTPBasic, HTTPBasicCredentials
from fastapi.security.utils import get_authorization_scheme_param
from starlette.requests import Request


class UTF8HTTPBasic(HTTPBasic):
    """Modified `fastapi.security.http.HTTPBasic` code."""

    async def __call__(  # type: ignore
        self, request: Request
    ) -> Optional[HTTPBasicCredentials]:
        authorization: str = request.headers.get("Authorization")
        scheme, param = get_authorization_scheme_param(authorization)
        unauthorized_headers = {"WWW-Authenticate": "Basic"}
        invalid_user_credentials_exc = HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid authentication credentials",
            headers=unauthorized_headers,
        )
        if not authorization or scheme.lower() != "basic":
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Not authenticated",
                headers=unauthorized_headers,
            )

        try:
            data = b64decode(param).decode("utf-8")
        except ValueError:
            raise invalid_user_credentials_exc

        username, separator, password = data.partition(":")
        if not separator:
            raise invalid_user_credentials_exc

        return HTTPBasicCredentials(username=username, password=password)
```

**Comment by pawamoy** [comment]
Just encountered the same issue. My application was asking again and again my credentials and I didn't get why at first. While debugging, I indeed saw that ASCII is expected. Maybe an `encoding` parameter could be added to the `HTTPBasic.__init__` method? That would also impact the username though 😕

**Comment by peter-jansen-tfs** [comment]
I encountered this issue yesterday.  my scenario involves an OAuth2 client_credentials flow.  we are using Authlib on the client side and FastAPI (HTTPBasic) on the server side.  in reviewing the relevant code in both packages, Authlib performs a URL quote on client_id and client_secret prior to the b64 encode.  as shown above, HTTPBasic only performs the b64 decode and no unquote.
[RFC6749 section 2.3.1](https://www.rfc-editor.org/rfc/rfc6749.html#page-16) states that the username and password must each be encoded using the "application/x-www-form-urlencoded" encoding algorithm .
since, in my scenario, the username and password are only referenced once, i perform a URL unquote on both prior to the reference.  this resolved the issue.
